File: assignments/go/bank-account/bank-account_test.go
package account

import "testing"

// The BankAccount module should define a type Account and support four calls:
//
// func OpenBank() : Account
//   Called at the start of each test. Returns an account handle.
//
// func CloseBank(account Account)
//   Called at the end of each test.
//
// func Balance(account Account) : int
//   Get the balance of the bank account.
//
// func Update(account Account, amount int)
//   Increment the balance of the bank account by the given amount.
//   The amount may be negative for a withdrawal.
//
// There should be a type Account which is returned from
//
// The initial value of the bank account should be 0.
//
// Remember that one of the guidelines in Effective Go is "share by
// communicating". Please assume in your design considerations that this is just
// a first prototype of an account system and that the final product will be
// much more complicated.

func assertBalance(t *testing.T, account Account, expected int) {
	if got := Balance(account); got != expected {
		t.Errorf("Balance(\"%v\"): Got \"%v\", expected \"%v\"", account, got, expected)
	}
}

func TestInitialBalanceZero(t *testing.T) {
	acc := OpenBank()
	defer CloseBank(acc)
	assertBalance(t, acc, 0)
}

func TestUpdateFromSameGoroutine(t *testing.T) {
	acc := OpenBank()
	defer CloseBank(acc)
	assertBalance(t, acc, 0)
	Update(acc, 10)
	assertBalance(t, acc, 10)
}

func TestUpdateFromDifferentGoroutine(t *testing.T) {
	acc := OpenBank()
	defer CloseBank(acc)
	assertBalance(t, acc, 0)
	c := make(chan int, 0)
	go func() {
		Update(acc, 10)
		c <- 1
	}()
	<-c
	assertBalance(t, acc, 10)
}

File: assignments/elixir/bank-account/bank_account_test.exs
if System.get_env("EXERCISM_TEST_EXAMPLES") do
  Code.load_file("example.exs")
else
  Code.load_file("account.exs")
end

ExUnit.start

# The BankAccount module should support four calls:
#
# open_bank()
#   Called at the start of each test. Returns an account handle.
#
# close_bank(account)
#   Called at the end of each test.
#
# balance(account)
#   Get the balance of the bank account.
#
# update(account, amount)
#   Increment the balance of the bank account by the given amount.
#   The amount may be negative for a withdrawal.
#
# The initial value of the bank account should be 0.

defmodule BankAccountTest do
  use ExUnit.Case, async: false # Tests should not overlap in execution.
  doctest BankAccount

  setup do
    account = BankAccount.open_bank()
    { :ok, [ account: account ] }
  end

  teardown context do
    BankAccount.close_bank(context[:account])
    :ok
  end

  test "initial balance is 0", context do
    assert BankAccount.balance(context[:account]) == 0
  end

  test "incrementing and checking balance", context do
    assert BankAccount.balance(context[:account]) == 0
    BankAccount.update(context[:account], 10)
    assert BankAccount.balance(context[:account]) == 10
  end

  test "incrementing balance from another process then checking it from test process", context do
    assert BankAccount.balance(context[:account]) == 0
    this = self()
    Process.spawn(fn ->
      BankAccount.update(context[:account], 20)
      send(this, :continue)
    end)
    receive do
      :continue -> :ok
    after 
      1000 -> flunk("Timeout") 
    end
    assert BankAccount.balance(context[:account]) == 20
  end

  ## Workarounds

  # This deals with the deprecation of the send operator.
  # It makes this code work both pre-0.12.2 and post-0.12.3
  unless { :send, 2 } in Kernel.__info__(:functions) do
    defp send(to, what), do: to <- what
  end
end
